File: R/Regression/Polynomial_Linear_Regression.R
# Polynomial Regression

# Import datasets
dataset = read.csv('Position_SalariesR.csv')
dataset = dataset[, 2:3] # Filtrar el dataset

# Adjust linear regression model
lin_reg = lm(formula = Dependent_var ~ .,
             data = dataset)

# Adjust polynomial regression model
dataset$ind_var2 = dataset$ind_var^2
dataset$ind_var3 = dataset$ind_var^3
dataset$ind_var4 = dataset$ind_var^4
poly_reg = lm(formula = Dependent_var ~ .,
              data = dataset)

# Linear model graph
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$ind_var, y = dataset$Dependent_var),
             color = "red") +
  geom_line(aes(x = dataset$ind_var, y = predict(lin_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("Linear model") +
  xlab("ind_var") +
  ylab("Dependent_var")


# Polynomial model graph
library(ggplot2)
x_grid = seq(min(dataset$ind_var), max(dataset$ind_var), 0.1)
ggplot() +
  geom_point(aes(x = dataset$ind_var, y = dataset$Dependent_var),
             color = "red") +
  geom_line(aes(x = x_grid, y = predict(poly_reg,
                                        newdata = data.frame(ind_var = x_grid,
                                                             ind_var2 = x_grid^2,
                                                             ind_var3 = x_grid^3,
                                                             ind_var4 = x_grid^4))),
            color = "blue") +
  ggtitle("Polynomial model") +
  xlab("ind_var") +
  ylab("Dependent_var")

# Linear model predictions
y_pred = predict(lin_reg, newdata = data.frame(ind_var = 137))

# Polynomial model predictions
y_PolyPred = predict(poly_reg, newdata = data.frame(ind_var = 137,
                                                    ind_var2 = 137^2,
                                                    ind_var3 = 137^3,
                                                    ind_var4 = 137^4))

File: Python/Regression/Polynomial_Linear_Regression.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('dataset.csv')
x = dataset.iloc[:, 1:2].values 
y = dataset.iloc[:, 2].values

# Divide dataset into training and testing sets
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

# Adjust regression model to data
from sklearn.linear_model import LinearRegression
linear_reg = LinearRegression()
linear_reg.fit(x, y)

# Adjust polynomial regression to dataset
from sklearn.preprocessing import PolynomialFeatures
pol_reg = PolynomialFeatures(degree = 4) # You must check for the best fitting degree
x_pol = pol_reg.fit_transform(x)
lin_reg_2 = LinearRegression() 
lin_reg_2.fit(x_pol, y)

# Plotting linear model
plt.scatter(x, y, color = 'red')
plt.plot(x, linear_reg.predict(x), color = 'blue')
plt.title("Simple linear regression model")
plt.xlabel("independent_var")
plt.ylabel("dependent_var")
plt.show()

# Plotting polynomial model
x_grid = np.arange(min(x), max(x), 0.1)  # For smoother curves
x_grid = x_grid.reshape(len(x_grid), 1)
plt.scatter(x, y, color = 'red')
plt.plot(x_grid, lin_reg_2.predict(pol_reg.fit_transform(x_grid)), color = 'blue') # plt.plot(x, lin_reg_2.predict(x_pol), color = 'blue') could also work for this
plt.title("Simple polynomial regression model")
plt.xlabel("independent_var")
plt.ylabel("dependent_var")
plt.show()

# Predictions from models (example value = 137)
linear_reg.predict([[137]]) # Linear model
lin_reg_2.predict(pol_reg.fit_transform([[137]])) # Polynomial model
